src/Gui/Inventor/Draggers/Gizmo.cpp

```diff
@@ -0,0 +1,697 @@
+}
+
+void RotationGizmo::placeBelowLinearGizmo(LinearGizmo* gizmo)
+{
+    linearGizmo = gizmo;
+
+    GizmoPlacement placement = gizmo->getDraggerPlacement();
+
+    draggerContainer->translation = Base::convertTo<SbVec3f>(placement.pos);
+    draggerContainer->setPointerDirection(-placement.dir);
+
+    translationSensor.setData(this);
+    translationSensor.setFunction(translationSensorCB);
+    translationSensor.setPriority(0);
+    SoSFVec3f& translation = gizmo->getDraggerContainer()->getDragger()->translation;
+    translationSensor.attach(&translation);
+    translation.touch();
+}
+
+double RotationGizmo::getRotAngle()
+{
+    double rotAngle = dragger->rotationIncrementCount.getValue()
+        * dragger->rotationIncrement.getValue();
+
+    return (rotAngle - addFactor) / multFactor;
+}
+
+void RotationGizmo::setRotAngle(double angle)
+{
+    angle = multFactor * angle + addFactor;
+    dragger->rotation = SbRotation({0, 0, 1.0f}, static_cast<float>(angle));
+}
+
+void RotationGizmo::setGeometryScale(float scale)
+{
+    dragger->geometryScale = SbVec3f(scale, scale, scale);
+}
+
+SoRotationDraggerContainer* RotationGizmo::getDraggerContainer()
+{
+    assert(draggerContainer);
+    return draggerContainer;
+}
+
+void RotationGizmo::draggingStarted()
+{
+    initialValue = property->value().getValue();
+    dragger->rotationIncrementCount.setValue(0);
+
+    if (delayedUpdateEnabled()) {
+        property->blockSignals(true);
+    }
+}
+
+void RotationGizmo::draggingFinished()
+{
+    if (delayedUpdateEnabled()) {
+        property->blockSignals(false);
+        property->valueChanged(property->value().getValue());
+    }
+}
+
+double normalizeIn360(double value)
+{
+    value = std::fmod(value, 360.0);
+    if (value < 0.0) {
+        value += 360.0;
+    }
+
+    return value;
+}
+
+double angularDist(double v1, double v2)
+{
+    return std::min(std::fabs(v1 - v2), 360 - std::fabs(v1 - v2));
+}
+
+// Returns a value between [0, 360) or (-180, 180] depending on if the
+// minimum value was positive or negetive. This is done because the taper angle
+// values in FreeCAD usually treat values like -10 and 350 differently
+double clampAngle(double value, double min, double max)
+{
+    value = normalizeIn360(value);
+    double nMin = normalizeIn360(min);
+    double nMax = normalizeIn360(max);
+
+    if (std::abs(nMax - nMin) > Base::Precision::Confusion()) {
+        if (nMax > nMin) {
+            if (value < nMin || value > nMax) {
+                value = angularDist(value, nMin) > angularDist(value, nMax)? nMax : nMin;
+            }
+        } else {
+            if (value < nMin && value > nMax) {
+                value = angularDist(value, nMin) > angularDist(value, nMax)? nMax : nMin;
+            }
+        }
+    }
+
+    if (min >= 0.0) {
+        // Return in [0, 360)
+        return value;
+    }
+
+    // Map to (-180, 180]
+    if (value > 180.0) {
+        value = value - 360;
+    }
+    return value;
+}
```

Comment:
could be moved to `Base/Tools.h` where other tools like that exist

Reply:
Should be fixed.